### ISSUE TITLE
#[deprecated] fn using format_args!() triggers "use of deprecated item" warning

### ISSUE THREAD

**Issue by shahn** [rust-lang/rust#35128]
This code:

    #[deprecated]
    fn foo() {
        format_args!("bla");
    }

gives this warning:

    warning: use of deprecated item, #[warn(deprecated)] on by default
     --> <anon>:3:18
    3 |>     format_args!("bla");
      |>                  ^^^^^
    <anon>:3:5: 3:25: note: in this expansion of format_args!

in current stable and nightly

**Comment by cengizIO** [comment]
Hello.

What's the real issue here? Are you concerned about the extra error message?

Because when I [try this in playground](https://is.gd/VuCh7z), I get two errors and one of them is correct behaviour.

```
warning: use of deprecated item, #[warn(deprecated)] on by default
 --> <anon>:3:18
3 |>     format_args!("bla");
  |>                  ^^^^^
<anon>:3:5: 3:25: note: in this expansion of format_args!

warning: use of deprecated item, #[warn(deprecated)] on by default
 --> <anon>:7:5
7 |>     foo();
  |>     ^^^
```

If it's so, I think you should update the title to "Deprecation warning getting incorrectly recursed into std function calls" or something like that.

Cheers!

**Comment by TimNN** [comment]
@cengizIO: I believe the current title is pretty accurate:

* only the `format_args!` macro is affected (as far as I can tell, `stringify!`, `concat!` and `try!` don't work for example)
* the error is triggered, even if `foo` is never used

See: https://is.gd/ua1Lyz

**Comment by cengizIO** [comment]
@TimNN Thanks for clarifying that for me. 

I traced source to [librustc_lint/builtin.rs#L577](https://github.com/rust-lang/rust/blob/master/src/librustc_lint/builtin.rs#L577) but can't seem to find something special for macros.

**Comment by TimNN** [comment]
Alright, I did some more tests and basically the following happens:

* `#[deprecated]` is applied recursively, so a `#[deprecated] mod foo {}` will deprecate anything inside that module as well.
* `format_args!` (probably) expands (among other things) to something like `const BLA: &'static str = "bla"; ...; do_somthing(BLA);`
* since `foo` is deprecated, `BLA` is deprecated, so using `BLA` in `do_something(BLA)` triggers a deprecation warning

Example:

```rust
#[deprecated]
fn _foo() {
    static FOO: &'static str = "abc";
    
    let _ = FOO; //~ WARN use of deprecated item
}
```

While this specific case is indeed very surprising, the more general case ("deprecated use inside of deprecated item") is already tracked at #16490.

**Comment by cengizIO** [comment]
@TimNN Nice catch! Here's the full expanded version

```
#![feature(prelude_import)]
#![no_std]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std as std;
#[deprecated]
fn _foo() {
    ::std::fmt::Arguments::new_v1({
                                      static __STATIC_FMTSTR: &'static [&'static str] = &["bla"];
                                      __STATIC_FMTSTR
                                  },
                                  &match () {
                                      () => [],
                                  });
}

fn main() {}
```
